federation-2/router-bridge/js-src/plan.ts

```diff
@@ -21,63 +47,463 @@ export class BridgeQueryPlanner {
+        },
+        errors: [e],
+      };
     }
+
+    // Adapted from here
+    // https://github.com/apollographql/apollo-server/blob/444c403011209023b5d3b5162b8fb81991046b23/packages/apollo-server-core/src/requestPipeline.ts#L303
+    const operationName = operation?.name;
+
+    // I double checked, this function doesn't throw
+    const operationDerivedData = getOperationDerivedData({
+      schema: this.apiSchema,
+      document,
+      operationName,
+    });
+
+    const statsReportKey = `# ${operationName || "-"}\n${
+      operationDerivedData.signature
+    }`;
+
+    return {
+      usageReporting: {
+        statsReportKey,
+        referencedFieldsByType: operationDerivedData.referencedFieldsByType,
+      },
+      data: this.planner.buildQueryPlan(operation),
+    };
   }
 }
 
 export function queryPlanner(schemaString: string): BridgeQueryPlanner {
   return new BridgeQueryPlanner(schemaString);
 }
 
-export function plan(
-  schemaString: string,
-  operationString: string,
-  operationName?: string
-): ExecutionResult<QueryPlan> {
-  try {
-    const composedSchema = buildSchema(schemaString);
-    const apiSchema = composedSchema.toAPISchema();
-    const operationDocument = parse(operationString);
-    const graphqlJsSchema = apiSchema.toGraphQLJSSchema();
+// ---------------------
 
-    // Federation does some validation, but not all.  We need to do
-    // all default validations that are provided by GraphQL.
-    const validationErrors = validate(graphqlJsSchema, operationDocument);
-    if (validationErrors.length > 0) {
-      return { errors: validationErrors };
-    }
+// Interface definition copied from here
+// https://github.com/apollographql/apollo-server/blob/d75c6cf3360a46ebcd944b2113438be8f549ae6f/packages/apollo-server-core/src/plugin/usageReporting/operationDerivedDataCache.ts#L5
+export interface OperationDerivedData {
+  signature: string;
+  referencedFieldsByType: ReferencedFieldsByType;
+}
+
+function getOperationDerivedData({
+  schema,
+  document,
+  operationName,
+}: {
+  schema: GraphQLSchema;
+  document: DocumentNode;
+  operationName: string;
+}): OperationDerivedData {
+  const generatedSignature = defaultUsageReportingSignature(
+    document,
+    operationName || ""
+  );
+
+  const generatedOperationDerivedData: OperationDerivedData = {
+    signature: generatedSignature,
+    referencedFieldsByType: calculateReferencedFieldsByType({
+      document,
+      schema,
+      resolvedOperationName: operationName ?? null,
+    }),
+  };
+  return generatedOperationDerivedData;
+}
 
-    const operation = operationFromDocument(
-      composedSchema,
-      operationDocument,
-      operationName
+// ---------------------
+
+// Adapted from here
+// https://github.com/apollographql/apollo-server/blob/06dd1171562564f0ca180e74a6fa5530e0389076/packages/apollo-server-core/src/plugin/usageReporting/referencedFields.ts
+
+export function calculateReferencedFieldsByType({
```

Comment:
shouldn't this be in query-planner-js instead, so it can be shared with the gateway?

Reply:
Absolutely! this will be [refactored eventually](https://github.com/apollographql/federation-rs/pull/92#discussion_r855714172)